Add Step Counter sensor for iOS

todi1856 opened the pull request:
* Added Step Counter sensor implementation for iOS, as you see both C# code and ObjC code lives in package.
* Made InputDevice.ExecuteCommand virtual this allows handling IOCTL commands on C# side, this also enables users to implement their own devices without waiting for us.
* Added iOS settings for Motion Usage

For testing:
- When you'll produce XCode project, please add Privacy - Motion Usage Description with any message. At this moment I am not sure if Input Package should be adding this value. UPDATE: Added a setting in Input Settings, input system package will add necessary info to Info.plist, see screenshot below
- When you'll launch app and try to Enable Step Counter - a dialog will pop asking for permission to use Motion, accept it. And start walking, after 3-10 seconds, the stepCounter value should start changing
- If you'll answer No to Motion dialog, the device will still be enabled, but no events will be coming. Even if you restart the app, the dialog won't longer show up, you either need to go to App Settings and enable Motion there, or reinstall app. I think though this is a bug in input system, the device should stay disabled if there was an error during Enable operation (like in this case authorization denial), i'll look into that in separate PR.


Outsanding question:
* iOSStepCounter doesn't support sample frequency, so if you'll try to use that property, you'll get NotSupportedException Device 'iOSStepCounter:/iOSStepCounter' does not support querying sampling frequency. Is that user friendly, or I should simply return 0. My personal opinion, I think the exception should stay, as it explicitly tells you not to use samplinFrequency property, but that's my opinion 

![ProjectSettingsWindow000](https://user-images.githubusercontent.com/568752/102370147-6cb76e80-3fc5-11eb-89d9-c99433dd6b7b.png)

todi1856 committed:
Introduce iOSStepCounter

todi1856 committed:
Finalizing step counter support

todi1856 committed:
typo

todi1856 committed:
remove unneeded includes

todi1856 committed:
More fixes

todi1856 committed:
Add test for iOSStepCounter

todi1856 committed:
changelog

glumazy commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,145 @@
+            if (t == DisableDeviceCommand.Type)
+            {
+#if UNITY_EDITOR
+                m_Enabled = true;
```
should it be false here and on line 108 ?

todi1856 replied:
correct, nice find

glumazy commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,145 @@
+        }
+
+#if UNITY_EDITOR
+        private bool m_Enabled;
```
does it need false as default value ?

glumazy commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm:
```diff
@@ -0,0 +1,124 @@
+
+extern "C" int _iOSStepCounterDisable(int deviceId)
+{
+    s_Wrapper.Disable(deviceId);
```
is it worth to return kResultFailure in case deviceId is diferent or m_Pedometer is null in s_Wrapper ?

todi1856 replied:
Makes sense

glumazy reviewed (approved):
LGTM

todi1856 committed:
Fix up nonsense with authorization

todi1856 committed:
Return full info about authorization

renaldasrep reviewed (approved):
Step Counter seems to be working fine or at least as expected, my only concern is that if `Privacy - Motion Usage Description` is not set - app crashes on device enable, would be nice to get a warning instead of a hard crash

Another issue Tomas thought of when we were discussing this is that the permission dialog runs on a separate thread - what if we enable multiple devices that require such dialog at once? It might not be an issue right now, but should be thought about and dealt with for the future

Last thing I noticed is that the device has 'samplingFrequency' property that's missing a getter. My test project reflects and prints all properties and it was not happy to find that, I had to add an if/else to bypass this. Might not be a huge issue for average user, but we should probably do something about this. Return a default value and spam a warning in the console 'samplingFrequency is returning 0 and this is expected, but please don't use this' - along those lines maybe?

todi1856 committed:
Add iOS specific settings

todi1856 committed:
Correctly update Info.plist

todi1856 committed:
Don't crash if motion usage description is missing in info.plist

todi1856 commented:
> Step Counter seems to be working fine or at least as expected, my only concern is that if `Privacy - Motion Usage Description` is not set - app crashes on device enable, would be nice to get a warning instead of a hard crash

Fixed that, it will now print to log instead of crashing

> 
> Another issue Tomas thought of when we were discussing this is that the permission dialog runs on a separate thread - what if we enable multiple devices that require such dialog at once? It might not be an issue right now, but should be thought about and dealt with for the future

That's something for later

> Last thing I noticed is that the device has 'samplingFrequency' property that's missing a getter. My test project reflects and prints all properties and it was not happy to find that, I had to add an if/else to bypass this. Might not be a huge issue for average user, but we should probably do something about this. Return a default value and spam a warning in the console 'samplingFrequency is returning 0 and this is expected, but please don't use this' - along those lines maybe?

Added a setting in Input Settings, where you can enable Motion Usage, and the relevent info will be added to Info.plist

todi1856 committed:
Expand iOSStepCounter test

renaldasrep commented:
The Motion Usage toggle+field is a very welcome addition, now it's way better. And doesn't crash!

todi1856 committed:
Fix compile errors

todi1856 committed:
iOSSupport.Initialize is called both for NativeInputRuntime and InputTestRuntime, but we only want to add real devices for NativeInputRuntime

jimon commented on Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsProvider.cs:
```diff
@@ -126,11 +126,20 @@ public override void OnGUI(string searchContext)
 
                 m_SupportedDevices.DoLayoutList();
 
+                EditorGUILayout.LabelField("iOS", EditorStyles.boldLabel);
```
A bit dunno tbh, is it worth putting platform specific detail into input system settings panel versus placing it into player settings where we already have established platform specific settings pattern? If I would be a user I would expect something changing the plist to be in the player settings, not input system.

Not to stop the PR, just want clarify what UX we expect here.

todi1856 replied:
We don't put package settings in Player Settings, as far as I am aware.

jimon commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+        [MonoPInvokeCallback(typeof(OnDataReceivedDelegate))]
+        private static void OnDataReceived(int deviceId, int numberOfSteps)
+        {
+            var stepCounter = (iOSStepCounter)InputSystem.GetDeviceById(deviceId);
```
I guess input system native API is not reachable from Obj-C? Otherwise it would probably be easier to just go directly to it instead of dealing with native->managed->native double transition.

todi1856 replied:
Yes, it's not reachable, but I don't understand where you see the last entry to native, you mean  InputSystem.GetDeviceById goes into native?

jimon replied:
- Your Obj-C code invokes manage code
- Your managed code queues event, which invokes native code

todi1856 replied:
Ah, I see... Thanks for explanation

jimon commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm:
```diff
@@ -0,0 +1,152 @@
+        m_Device = deviceId;
+        [m_Pedometer startPedometerUpdatesFromDate: [NSDate date] withHandler:^(CMPedometerData * _Nullable pedometerData, NSError * _Nullable error)
+        {
+            // Note: We need to call our callback on the same thread the Unity scripting is operating
```
If we could get access to input module native api here, then no, it can be queued from any thread

todi1856 replied:
I'll try to explore, since dispatching like this - probably delays the input event by quite a lot

jimon commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm:
```diff
@@ -0,0 +1,152 @@
+
+#define ENABLE_STEP_COUNTER_LOGGING 1
+#if ENABLE_STEP_COUNTER_LOGGING
+#define STEP_COUNTER_LOG(...) NSLog(@"StepCounter - %@", [NSString stringWithFormat: __VA_ARGS__])
```
I don't see any code disabling that, meaning we spam logs in release builds?

todi1856 replied:
It's only logging in case of errors

jimon replied:
Ah ok, missed that

jimon commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.mm:
```diff
@@ -0,0 +1,152 @@
+    if (@available(iOS 11.0, *))
+    {
+        return (int)[CMPedometer authorizationStatus];
+    }
```
You could also use isAuthorizedForRecording for <11.0 here

todi1856 replied:
Didn't know that exist, thanks

todi1856 replied:
Actually there's no such function for CMPedometer

jimon commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSSupport.cs:
```diff
@@ -41,6 +43,16 @@ public static void Initialize()
+
+            // Don't add devices for InputTestRuntime
+            // TODO: Maybe there should be a better place for adding device from C#
+            if (InputSystem.s_Manager.m_Runtime is NativeInputRuntime)
```
Yea it seems suboptimal to do jumpyloops with checking which runtime we're using ... not quite sure what would be a better way to do it tbh.

jimon reviewed (approved):
It's ok PR given the current limitations of input system. A better way IMHO would to avoid native->managed->native double transition on every pedometer event. Also would be great to change input native module API somewhat to accommodate a case of registering "singleton" device better.

Rene-Damm commented on Packages/com.unity.inputsystem/CHANGELOG.md:
````diff
@@ -52,6 +52,7 @@ however, it has to be formatted properly to pass verification tests.
                   Debug.Log($"Button {button} was pressed");
       }
   ```
+  - Added Step Counter sensor for iOS, you'll need to add 'Privacy - Motion Usage Description' to Info.plist before using the sensor.
````
```
- Added support for step counters on iOS.
  ``
  InputSystem.EnableDevice(iOSStepCounter.current);
  ``
  * You need to add `Privacy - Motion Usage Description` to `Info.plist` to be able to use the sensor in your app.
```

todi1856 replied:
Changed the changelog since, this entry was before I did iOS input settings

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/InputSettingsiOS.cs:
```diff
@@ -0,0 +1,34 @@
+    public partial class InputSettings : ScriptableObject
+    {
+        [Serializable]
+        public class iOSSettings
```
Needs API docs.

todi1856 replied:
Done

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/InputSettingsiOSProvider.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace UnityEngine.InputSystem
+{
+    public class InputSettingsiOSProvider
```
Does this need to be public. If so, needs to be documented.

todi1856 replied:
made internal, also caught by tests

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSPostProcessBuild.cs:
```diff
@@ -0,0 +1,35 @@
+using UnityEngine;
+using UnityEngine.InputSystem;
+
+public class iOSPostProcessBuild
```
Shouldn't be in global namespace. Can this be internal? If it has to be public, needs to be documented.

From the looks of it, seems like this should be internal.

todi1856 replied:
Fixed, also made internal

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+namespace UnityEngine.InputSystem.iOS.LowLevel
+{
+    //See CoreMotion.framework/Headers/CMAuthorization.h
+    public enum MotionAuthorizationStatus : int
```
Xmldocs

todi1856 replied:
fixed

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsProvider.cs:
```diff
@@ -126,11 +126,20 @@ public override void OnGUI(string searchContext)
+        private static void ShowPlatformSettings()
+        {
+            // Would be nice to get BuildTargetDiscovery.GetBuildTargetInfoList since that contains information about icons etc
+        }
```
Dead code? Add `////TODO` or `////REVIEW` for things that should be changed/considered in the future.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/InputSettingsiOS.cs:
```diff
@@ -0,0 +1,34 @@
+#if UNITY_EDITOR || UNITY_IOS || UNITY_TVOS
```
Needs `PACKAGE_DOCS_GENERATION` in the #if.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+#if UNITY_EDITOR || UNITY_IOS || UNITY_TVOS
```
Probably needs `PACKAGE_DOCS_GENERATION` in the #if.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSSupport.cs:
```diff
@@ -41,6 +43,16 @@ public static void Initialize()
+            if (InputSystem.s_Manager.m_Runtime is NativeInputRuntime)
+            {
+                if (iOSStepCounter.IsAvailable())
+                    InputSystem.AddDevice<iOSStepCounter>();
```
How come we're doing this one differently and adding it from managed instead of reporting it from native?

todi1856 replied:
I believe, we always had the option to have everything in input system package... sadly I went with general flow in the beginning and did like other platforms did - half of the code in Unity core, other half in input system package.

I think the approach where you have the whole implementation in input system package, gives tremendous flexibility (same as screen keyboard impl) - we don't need 2 PRs to fix something, users are able to see implementation, it also shows how users from their project can add their own devices without being blocked by us.

Rene-Damm replied:
>I think the approach where you have the whole implementation in input system package, gives tremendous flexibility

Yup, agree.

I think should still report by device description, though. Doing `AddDevice<XXX>`, we bypass the layout matching process and thus don't give an opportunity for someone rewire the layout matching.

Rene-Damm replied:
Indeed needs work on our end. Created a [ticket](https://jira.unity3d.com/browse/ISX-449) for it.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+        /// <summary>
+        /// Does the phone supports the pedometer?
+        /// </summary>
+        /// <returns></returns>
```
No empty stuff.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        /// <summary>
+        /// Does the phone supports the pedometer?
```
"support"

todi1856 replied:
fixed

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        /// <summary>
+        /// Query motion authorization status
```
Docs like this are so arghhhhhhhhhh. This is the stuff that makes Unity API docs really disheartening to visit.

If the doc comment doesn't doc anything, preferably leave it out altogether.

todi1856 replied:
Rephrased

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Plugins/iOS/iOSStepCounter.cs:
```diff
@@ -0,0 +1,166 @@
+    /// <summary>
+    /// Step Counter (also known as pedometer) sensor for iOS.
+    /// Note: You need to add 'Privacy - Motion Usage Description' to Info.plist to use this sensor.
+    /// </summary>
```
Could use a simple code example that shows that you need to enable the sensor through the API and how to do it.

todi1856 replied:
Fixed

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsProvider.cs:
```diff
@@ -301,6 +310,8 @@ private void InitializeWithCurrentSettings()
                 }
             };
+
+            m_iOSProvider = new InputSettingsiOSProvider(m_SettingsObject);
```
For now, this is probably okay. However, going forward, this won't work across platforms. E.g. with consoles, we can't do this. Preferably, this should be done using the new polymorphic serialization. So it'd just be a list of SerializableObjects. Would also be easy to make the UI extensible this way.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs:
```diff
@@ -519,7 +519,7 @@ protected virtual void OnRemoved()
         /// target="_blank">ioctl</a> on UNIX-like systems.
         /// </remarks>
-        public unsafe long ExecuteCommand<TCommand>(ref TCommand command)
+        public virtual unsafe long ExecuteCommand<TCommand>(ref TCommand command)
```
Hmm, I remember this one led to compilation errors on PS4 back then. Was an il2cpp issue, though, so may have been fixed. Just remember that the particular combination of virtual + parameterized + ref parameter blew up il2cpp. An easy workaround, if we still need it, is to just have a separate, non-parameterized version that is virtual and just takes a direct pointer instead.

todi1856 replied:
Adjusted the code as per your suggestion

todi1856 committed:
Merge remote-tracking branch 'origin/develop' into ios/step-counter

todi1856 committed:
Adjust CHANGELOG.md

todi1856 committed:
Fix up MotionUsage API

todi1856 committed:
docs

todi1856 committed:
Refactor ExecuteCommand

todi1856 committed:
Fix tooltip

todi1856 committed:
Fixing tests

todi1856 committed:
More informative message

todi1856 committed:
Make InputSettingsiOSProvider internal, no need for it to be public

todi1856 committed:
Fix docs

todi1856 committed:
Fix compile error

todi1856 committed:
Docs

todi1856 committed:
remove empty

todi1856 committed:
Fix Step Counter test on iOS

Commit:
MERGE: develop => ios/step-counter.

Commit:
DOCS: Doc and API tweaks.
